File: admin-backend/app/api/routes/applications.py
"""Application routes."""
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.api.deps import get_db_session, require_admin
from app.schemas.application import ApplicationCreate, ApplicationResponse, ApplicationListResponse
from app.services.application import ApplicationService

router = APIRouter()


def _to_application_response(app) -> ApplicationResponse:
    return ApplicationResponse(
        id=app.id,
        user_id=app.user_id,
        reason=app.reason,
        status=app.status,
        created_at=app.created_at,
        reviewed_at=app.reviewed_at,
        user_username=app.user.username if app.user else None,
        user_email=app.user.email if app.user else None,
    )


@router.post("", response_model=ApplicationResponse)
def create_application(
    data: ApplicationCreate,
    db: Annotated[Session, Depends(get_db_session)],
):
    """User submits API Key application."""
    svc = ApplicationService(db)
    try:
        app = svc.create_application(data)
        app = svc.get_application(app.id)
        return _to_application_response(app)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("", response_model=ApplicationListResponse)
def list_applications(
    db: Annotated[Session, Depends(get_db_session)],
    _: Annotated[None, Depends(require_admin)],
    status: Annotated[Optional[str], Query()] = None,
    skip: Annotated[int, Query(ge=0)] = 0,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
):
    """Admin: list all applications, optionally filter by status."""
    svc = ApplicationService(db)
    items, total = svc.list_applications(status=status, skip=skip, limit=limit)
    return ApplicationListResponse(
        items=[_to_application_response(a) for a in items],
        total=total,
    )


@router.post("/{application_id}/approve")
def approve_application(
    application_id: int,
    db: Annotated[Session, Depends(get_db_session)],
    _: Annotated[None, Depends(require_admin)],
):
    """Admin: approve application and create API Key + APISIX Consumer."""
    svc = ApplicationService(db)
    try:
        credential = svc.approve(application_id)
        return {
            "message": "Approved",
            "application_id": application_id,
            "credential_id": credential.id,
            "api_key": credential.api_key,
            "consumer_name": credential.consumer_name,
            "quota_per_minute": credential.quota_per_minute,
            "token_balance": credential.token_balance,
        }
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=502, detail=str(e))


@router.post("/{application_id}/reject")
def reject_application(
    application_id: int,
    db: Annotated[Session, Depends(get_db_session)],
    _: Annotated[None, Depends(require_admin)],
):
    """Admin: reject application."""
    svc = ApplicationService(db)
    try:
        app = svc.reject(application_id)
        return {"message": "Rejected", "application_id": app.id}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.delete("/{application_id}")
def delete_application(
    application_id: int,
    db: Annotated[Session, Depends(get_db_session)],
    _: Annotated[None, Depends(require_admin)],
):
    """Admin: permanently delete application (if approved, removes credential and APISIX consumer)."""
    svc = ApplicationService(db)
    try:
        svc.delete_application(application_id)
        return {"message": "Deleted", "application_id": application_id}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: frontend/src/api/applications.ts
import { ADMIN_API_BASE } from '../config'
import type {
  ApplicationCreate,
  Application,
  ApplicationListResponse,
  ApplicationStatusResponse,
  ApproveResponse,
} from '../types'
import { adminGet, adminPost, adminDelete, ApiError } from './client'

const BASE = '/api/v1/applications'

/** 用户：提交申请 */
export async function createApplication(
  data: ApplicationCreate,
  adminKey?: string | null
): Promise<Application> {
  try {
    return await adminPost<Application, ApplicationCreate>(BASE, data, adminKey)
  } catch (e) {
    throw e as ApiError
  }
}

/** 用户：凭申请编号 + 邮箱查询审批状态（已通过时返回 API Key） */
export async function getApplicationStatus(
  applicationId: number,
  email: string
): Promise<ApplicationStatusResponse> {
  const url = `${ADMIN_API_BASE}${BASE}/${applicationId}/status?email=${encodeURIComponent(email.trim())}`
  const res = await fetch(url)
  const data = await res.json().catch(() => ({}))
  if (!res.ok) {
    const err = new Error((data.detail ?? data.message ?? res.statusText) as string) as ApiError
    ;(err as { status?: number }).status = res.status
    throw err
  }
  return data as ApplicationStatusResponse
}

/** 管理员：获取申请列表 */
export async function listApplications(
  params: { status?: string; skip?: number; limit?: number } = {},
  adminKey?: string | null
): Promise<ApplicationListResponse> {
  const sp = new URLSearchParams()
  if (params.status) sp.set('status', params.status)
  if (params.skip != null) sp.set('skip', String(params.skip))
  if (params.limit != null) sp.set('limit', String(params.limit))
  const q = sp.toString()
  const path = q ? `${BASE}?${q}` : BASE
  try {
    return await adminGet<ApplicationListResponse>(path, adminKey)
  } catch (e) {
    throw e as ApiError
  }
}

/** 管理员：审批通过 */
export async function approveApplication(
  applicationId: number,
  adminKey?: string | null
): Promise<ApproveResponse> {
  try {
    return await adminPost<ApproveResponse>(
      `${BASE}/${applicationId}/approve`,
      undefined,
      adminKey
    )
  } catch (e) {
    throw e as ApiError
  }
}

/** 管理员：拒绝 */
export async function rejectApplication(
  applicationId: number,
  adminKey?: string | null
): Promise<{ message: string; application_id: number }> {
  try {
    return await adminPost<{ message: string; application_id: number }>(
      `${BASE}/${applicationId}/reject`,
      undefined,
      adminKey
    )
  } catch (e) {
    throw e as ApiError
  }
}

/** 管理员：永久删除申请（已通过的会同时删除 API Key 与 APISIX Consumer） */
export async function deleteApplication(
  applicationId: number,
  adminKey?: string | null
): Promise<{ message: string; application_id: number }> {
  try {
    return await adminDelete(`${BASE}/${applicationId}`, adminKey)
  } catch (e) {
    throw e as ApiError
  }
}
